analyzers/src/SonarAnalyzer.Common/Trackers/ArgumentTracker.cs

```diff
@@ -0,0 +1,86 @@
+﻿/*
+* SonarAnalyzer for .NET
+* Copyright (C) 2015-2023 SonarSource SA
+* mailto: contact AT sonarsource DOT com
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 3 of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this program; if not, write to the Free Software Foundation,
+* Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*/
+
+namespace SonarAnalyzer.Helpers.Trackers;
+
+public abstract class ArgumentTracker<TSyntaxKind> : SyntaxTrackerBase<TSyntaxKind, ArgumentContext>
+    where TSyntaxKind : struct
+{
+    protected abstract RefKind? ArgumentRefKind(SyntaxNode argumentNode);
+    protected abstract IReadOnlyCollection<SyntaxNode> ArgumentList(SyntaxNode argumentNode);
+    protected abstract int? Position(SyntaxNode argumentNode);
+    protected abstract bool InvocationFitsMemberKind(SyntaxNode invokedExpression, MemberKind memberKind);
+    protected abstract bool InvokedMemberFits(SemanticModel model, SyntaxNode invokedExpression, MemberKind memberKind, Func<string, bool> invokedMemberNameConstraint);
+
+    protected override ArgumentContext CreateContext(SonarSyntaxNodeReportingContext context) =>
+        new(context);
+
+    public Condition MatchArgument(ArgumentDescriptor descriptor) =>
+        context =>
+        {
+            if (context.Node is { } argumentNode
+                && argumentNode is { Parent.Parent: { } invoked }
+                && SyntacticChecks(context.SemanticModel, descriptor, argumentNode, invoked)
+                && (descriptor.InvokedMemberNodeConstraint?.Invoke(context.SemanticModel, Language, invoked) ?? true)
+                && MethodSymbol(context.SemanticModel, invoked) is { } methodSymbol
+                && Language.MethodParameterLookup(invoked, methodSymbol).TryGetSymbol(argumentNode, out var parameter)
+                && ParameterFits(parameter, descriptor.ParameterConstraint, descriptor.InvokedMemberConstraint))
+            {
+                context.Parameter = parameter;
```

Comment:
I find it very confusing that this method has side effects. Will this used only in tests?

Reply:
> Will this used only in tests?
Unfortunately not. The PR uses this call. 

> this method has side effects
Yes. This is very unfortunate. There should be a distinction between input and output, but the context is both at the moment. This is how the tracker is designed now: It creates a pipeline of matchers and passes the context from one to the others. Each matcher can manipulate parts of the context. (It's a bit like the request pipeline in asp.net core).